### PR TITLE
[PE-2246] feat: bubble-up video events to ix-video

### PR DESCRIPTION
**Pull request by luqven**
## Changes

- creates more granular event details for certain event listeners. 
	- This allows users to do more with the events that they are listening to
	- like calculating buffered % or determining the muted/unmuted status of the video.
- leverages VJS' `fluid` mode to have the video resize on
viewport resizing.
	  - video resizes when the viewport size changes
	  - video respects container size
	  - video maintains the aspect ratio on resize
	  - setting w/h without `fixed` attribute maintains a.r.
	  - setting w/h with `fixed` attr disregards a.r. and forces dims.
- refactors state so that modification is done at the last possible second of both state and properties
- makes use of state and converters to set the shape of properties as early and consistently as possible
- introduces an `updated()` hook that enables us to update properties in response to an attribute change.
- Centers the big play button

### Example Vue app where we use `@seeking` event listener and handler

```vue
<script>
import {IxVideo} from '@imgix/web-components'
import {ref} from 'vue'
// ...
const ixVideo = ref(null);
export default {
  methods: {
    warn(message, event) {
      const video = ixVideo.value
      console.info(video)
      console.info(message, event.detail)
    },
    toggleControls() {
      const video = ixVideo.value
      console.info(video.controls)
      video.controls = !video.controls
      console.info(video.controls)
      
    }
  }
}
</script>

<template>
  <main>
    <div
    class="mobile"
    style="display: flex; justify-content: center; margin: auto"
    width="600px"
    height="400px"
  >
  <button @click="() => toggleControls()" >Toggle Controls</button>

    <ix-video
      ref="ixVideo"
      controls
      source="https://assets.imgix.video/videos/girl-reading-book-in-library.mp4"
      data-setup='{ "playbackRates": [0.5, 1, 1.5, 2] }'
      @seeked="(event) => warn('seeked:', event)"
      @seeking="(event) => warn('seeking:', event)"
    ></ix-video>
  </div>
  </main>
</template>

```

https://user-images.githubusercontent.com/16711614/166343868-d65534bb-21f8-45f3-919c-5f0ee7f60894.mov

https://user-images.githubusercontent.com/16711614/166344759-0622a14d-5856-4349-8229-c775cf3dfc87.mov

https://user-images.githubusercontent.com/16711614/166345409-e4f5c1bd-60d7-437b-9b18-1c672c26e8d1.mov

### Stacked PR Chain: PE-2246
| PR | Title | Status |  Merges Into  |
|:--:|:------|:-------|:-------------:|
|#34|👉feat: bubble-up video events to ix-video||-|
|#35|~~refactor: ensure unset attrs not set to emtpy str~~||#34|
|#36|~~test: refactor tests into sep specs~~||#35|
|#37|~~test: ix-video react integration tests~~||#36|
|#39|~~feat: make video sizing responsive by default~~||#37|
|#38|~~feat: react to props change~~||#39|
|#40|~~test: ensure props change is reflected on video~~||#38|

### CODE REVIEW

File: src/elements/ix-video.ts [hunk 1]
```diff
@@ -166,17 +166,15 @@ export class IxVideo extends LitElement {
     listener: EventListenerOrEventListenerObject,
     options?: boolean | EventListenerOptions
   ) => {
-    const event = DefaultVideoEventsMap[type];
```
Comment (luqven) [inline src/elements/ix-video.ts]: The reason `DefaultVideoEventsMap` is a map and not just an array of strings is to 1) avoid misspellings and 2) allow us to cheapy check for the presence of event names in the map in the future to validate listeners added by users.

File: cypress/integration/ix-video/data-setup.spec.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    });
+
+    it('should display video controls', () => {
+      cy.get(bigPLayButton).should('exist');
+    });
+
+    it('should display large play button on first render', () => {
+      cy.get(bigPLayButton).should('exist');
+    });
+
```
Comment (sherwinski) [inline cypress/integration/ix-video/data-setup.spec.ts:38]: Just FYI, these two seem to be testing the same thing.

File: cypress/integration/ix-video/mp4-source.spec.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      cy.get(player).should('exist');
+    });
+
+    it('should display video controls', () => {
+      cy.get(bigPLayButton).should('exist');
+    });
+
+    it('should display large play button on first render', () => {
+      cy.get(bigPLayButton).should('exist');
+    });
```
Comment (sherwinski) [inline cypress/integration/ix-video/mp4-source.spec.ts:34]: Same here

File: cypress/integration/ix-video/react.spec.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    cy.get(player).should('exist');
+  });
+
+  it('should display video controls', () => {
+    cy.get(bigPLayButton).should('exist');
+  });
+
+  it('should display large play button on first render', () => {
+    cy.get(bigPLayButton).should('exist');
+  });
```
Comment (sherwinski) [inline cypress/integration/ix-video/react.spec.ts]: I suppose if there isn't a straightforward way to differentiate these tests then we should consider removing one?

**Review by sherwinski — APPROVED** [review]
🚀 🚀

File: cypress/integration/ix-video/buttons.spec.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  PLAY_BUTTON_SELECTOR,
+} from '../../fixtures/selectors';
+
+/**
+ * TODO:
+ * - Read the document w/h values and store them to compare against
+ * - Split each describe block into it's own test
+ * - Move selectors into fixture files
+ * - Add helper and conversion functions unit tests
+ */
```
[the remark this answers is not in the document]
Comment (luqven) [reply cypress/integration/ix-video/buttons.spec.ts:14]: image

File: cypress/integration/ix-video/data-setup.spec.ts (hunk 2 above)
Comment (luqven) [reply cypress/integration/ix-video/data-setup.spec.ts:38]: 💯 , meant to remove one from both of those tests apols.

File: cypress/integration/ix-video/buttons.spec.ts (hunk 5 above)
Comment (luqven) [inline cypress/integration/ix-video/buttons.spec.ts:14]:
```suggestion
```

File: cypress/integration/ix-video/data-setup.spec.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      cy.get(player).should('exist');
+    });
+
+    it('should display video controls', () => {
+      cy.get(bigPLayButton).should('exist');
+    });
+
+    it('should display large play button on first render', () => {
+      cy.get(bigPLayButton).should('exist');
+    });
```
Comment (luqven) [inline cypress/integration/ix-video/data-setup.spec.ts:37]:
```suggestion
```

File: cypress/integration/ix-video/mp4-source.spec.ts (hunk 3 above)
Comment (luqven) [inline cypress/integration/ix-video/mp4-source.spec.ts:34]:
```suggestion
```

File: cypress/integration/ix-video/react.spec.ts (hunk 4 above)
Comment (luqven) [inline cypress/integration/ix-video/react.spec.ts]:
```suggestion
```

**Review by luqven — COMMENTED** [review]
Going to go ahead and remove these duplicated tests ✅

File: cypress/integration/ix-video/react.spec.ts (hunk 4 above)
Comment (luqven) [inline cypress/integration/ix-video/react.spec.ts]:
```suggestion
```

File: cypress/integration/ix-video/styles.spec.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    cy.get(videoTag).should('exist');
+  });
+
+  it('should display video controls', () => {
+    cy.get(bigPLayButton).should('exist');
+  });
+
+  it('should display large play button on first render', () => {
+    cy.get(bigPLayButton).should('exist');
+  });
```
Comment (luqven) [inline cypress/integration/ix-video/styles.spec.ts:35]:
```suggestion
```

File: cypress/integration/ix-video/react.spec.ts (hunk 4 above)
Comment (luqven) [inline cypress/integration/ix-video/react.spec.ts]:
```suggestion
```

**Pull request merged**